PurchasesCoreSwift/SubscriberAttributes/SubscriberAttributesManager.swift

```diff
@@ -0,0 +1,263 @@
+        setAttribute(key: SpecialSubscriberAttributes.ip, value: "true", appUserID: appUserID)
+    }
+
+    @objc public func syncAttributesForAllUsers(currentAppUserID: String) {
+        let unsyncedAttributesForAllUsers = unsyncedAttributesByKeyForAllUsers()
+
+        for (syncingAppUserId, attributes) in unsyncedAttributesForAllUsers {
+            syncAttributes(attributes: attributes, appUserID: syncingAppUserId) { error in
+                self.handleAttributesSynced(syncingAppUserId: syncingAppUserId,
+                                            currentAppUserId: currentAppUserID,
+                                            error: error)
+            }
+        }
+    }
+
+    @objc public func handleAttributesSynced(syncingAppUserId: String, currentAppUserId: String, error: Error?) {
+        if error == nil {
+            Logger.rcSuccess(String(format: Strings.attribution.attributes_sync_success, syncingAppUserId))
+            if syncingAppUserId != currentAppUserId {
+                deviceCache.deleteAttributesIfSynced(appUserID: syncingAppUserId)
+            }
+        } else {
+            let receivedNSError = error as NSError?
+            Logger.error(String(format: Strings.attribution.attributes_sync_error,
+                                receivedNSError?.localizedDescription ?? "",
+                                receivedNSError?.userInfo ?? ""))
+        }
+    }
+
+    @objc public func unsyncedAttributesByKey(appUserID: String) -> SubscriberAttributeDict {
+        // TODO  these logs were in RCPurchases+subscriberattributes, are we ok with logging
+        // them every time?
+        let unsyncedAttributes = deviceCache.unsyncedAttributesByKey(appUserID: appUserID)
+        Logger.debug(String(format: Strings.attribution.unsynced_attributes_count,
+                            unsyncedAttributes.count,
+                            appUserID))
+        if unsyncedAttributes.count > 0 {
+            Logger.debug(String(format: Strings.attribution.unsynced_attributes, unsyncedAttributes))
+        }
+
+        return unsyncedAttributes
+    }
+
+    @objc public func unsyncedAttributesByKeyForAllUsers() -> [String: SubscriberAttributeDict] {
+        return deviceCache.unsyncedAttributesForAllUsers()
+    }
+
+    @objc public func markAttributesAsSynced(_ maybeAttributesToSync: SubscriberAttributeDict?, appUserID: String) {
+        guard let attributesToSync = maybeAttributesToSync,
+              !attributesToSync.isEmpty else {
+            return
+        }
+
+        Logger.info(String(format: Strings.attribution.marking_attributes_synced,
+                           appUserID,
+                           attributesToSync.description))
+
+        lock.lock()
+        var unsyncedAttributes = unsyncedAttributesByKey(appUserID: appUserID)
+        for (key, attribute) in attributesToSync {
+            if let unsyncedAttribute = unsyncedAttributes[key] {
+                if unsyncedAttribute == attribute {
```

Comment:
this is a departure from previous behavior: 
This is checking the `SubscriberAttribute` instances, which define equality as [key, value, setTime and synced](https://github.com/RevenueCat/purchases-ios/blob/swift_migration/PurchasesCoreSwift/SubscriberAttributes/SubscriberAttribute.swift#L59). 

this means that if the value for `isSynced` was different, or if setTime is different, it won't mark them as synced. 

Old behavior is to [check attribute.value](https://github.com/RevenueCat/purchases-ios/blob/swift_migration/Purchases/SubscriberAttributes/RCSubscriberAttributesManager.m#L212), I'd go with that to be conservative.

Reply:
eeeek good catch